### DISCUSSION TITLE
Using GitHub Packages with GitHub Apps

### DISCUSSION THREAD

**Discussion by hamuyuuki** [community/community#26920] [Actions]
<p>I’d like to manage a private RubyGems on GitHub Packages. And I’d like to use a personal access token or GitHub Apps’s token(installation access token) to authenticate.</p>
<p>I could use a personal access token to authenticate. But i couldn’t use GitHub Apps’s token(installation access token). I explain here how to try.</p>
<ol>
<li>Create GitHub Apps with  <em>Packages</em> (read and write access level) of Repository permissions.</li>
<li>Install GitHub Apps to a repository that managing a private RubyGems.</li>
<li>Generate GitHub Apps’s token(installation access token) with a private key.</li>
</ol>
<ul>
<li>ref: <a href="https://docs.github.com/en/developers/apps/authenticating-with-github-apps#authenticating-as-an-installation" rel="nofollow noopener">https://docs.github.com/en/developers/apps/authenticating-with-github-apps#authenticating-as-an-installation</a>
</li>
</ul>
<ol start="4">
<li>Fetch a private RubyGems from a repository with GitHub Apps’s token(installation access token).</li>
</ol>
<p>GitHub Docs explains to use GitHub Packages with a personal access token to authenticate. But I didn’t find about using GitHub Packages with GitHub Apps.</p>
<p>Can i use GitHub Packages with GitHub Apps now?</p>
<p>Thank you.</p>

**Comment by jcansdale** [comment, marked as answer]
<aside class="quote no-group" data-username="hamuyuuki" data-post="1" data-topic="130950">
<div class="title">
<div class="quote-controls"></div>
 hamuyuuki:</div>
<blockquote>
<p>Can i use GitHub Packages with GitHub Apps now?</p>
</blockquote>
</aside>
<p>GitHub Packages doesn’t currently support GitHub App tokens. The only exception is the default GitHub Actions token, which is a special kind of app token. For the moment, you will need to use a PAT for programmatic access.</p>
<p>Regards,<br>
Jamie.</p>

**Reply by renan-alm** [reply]
Quick question on this. This was the case in 2020, do we have this support nowadays (2024)?

**Reply by seshasandeep** [reply]
Same question as above :)

**Reply by konstantinj** [reply]
Still seems to be an issue, at least for me.

**Reply by perok** [reply]
What does the "packages" permissions actually allow per now given that it _can't_ access packages? :thinking:

**Reply by Omar-Gebal** [reply]
@renan-alm Did you happen to find the answer to that ?

**Reply by renan-alm** [reply]
@Omar-Gebal no change. [Packages only supports authentication using a PAT](https://docs.github.com/en/packages/learn-github-packages/about-permissions-for-github-packages#about-scopes-and-permissions-for-package-registries) sadly.

**Reply by Omar-Gebal** [reply]
@renan-alm RIP, thanks for the reply, Also I found [this](https://github.com/orgs/community/discussions/36441) if you're still curious. Looks like they will add support for it but it still isn't implemented
<img width="639" height="455" alt="image" src="https://github.com/user-attachments/assets/23014c6c-cedf-416a-9393-7cb1c8d44ee4" />

**Reply by Radiergummi** [reply]
Oh my. This is so disappointing. Using PATs is such a bad option for companies - employee-bound tokens embedded in IaC or Actions secrets are nothing but ticking time-bombs. At some point an employee is going to leave, or worse, has to be fired due to trust issues.  
Now you're in for a hunt: Did you really replace all tokens? Will any infrastructure component fail due to an unexpected *Access Denied* error?  
The alternative is to have a bot account exclusively used to hold personal access tokens, but having to pay for an extra seat just to cover for GitHub's awful product choices and forever-dangling roadmap items makes me *really* angry.

**Comment by hamuyuuki** [comment]
<p>Thank you for answering. I understand :smiley:</p>

**Comment by singhprd** [comment]
It would be great if adding package support to apps was considered as an improvement.

At the moment to get packages access to a ruby gem for example you need:
- a GitHub user
- a Personal Access Token (classic) - [fine grained is not supported yet](https://github.com/orgs/community/discussions/36441#discussioncomment-3914371)

Allowing a GitHub app to create tokens for package access would mean you:
- don't need a machine user/to overload a user with a PAT
- could cycle the access token programmatically 

Thanks!

**Comment by stekern** [comment]
Any updates to this? This (and [GitHub Packages support for fine-grained PATs](https://github.com/github/roadmap/issues/558), which I assume is related) is sorely needed functionality. 

GitHub applications like [Renovate](https://github.com/renovatebot/renovate) are blocked from working seamlessly with GitHub Packages and today require the use of user-associated, classic personal access tokens to fetch package information from GitHub Packages.

**Reply by AllexVeldman** [reply]
I was able to get Renovate to use a GITHUB_TOKEN to authenticate to ghcr.io.
It does require you to self-host Renovate.

I wrote a small description as part of a sideproject of mine: https://github.com/AllexVeldman/pyoci?tab=readme-ov-file#renovate--ghcrio

**Comment by behluluysal** [comment]
Is PATs still the only option to use GitHub Packages?

**Comment by azlam-abdulsalam** [comment]
Any updates on this?

**Comment by enniobolondi** [comment]
Any news?

**Comment by russovincenzo** [comment]
It's a bug or not? Why deny access to packages with app token ?

**Comment by joluizquierdo** [comment]
I tested it myself today, GitHub App installation tokens aren't supported, can you please update on this?
![image](https://github.com/user-attachments/assets/b9f25c41-c9f1-4929-9c4a-72d740080cf7)

Thank you!!!

**Comment by 238alexg** [comment]
Similar to folks above, I was very confused by this. I was swapping my CICD from using my own personal access token to use GitHub App tokens. PAT obviously isn't great for CICD because it requires sharing personal credentials.

The messaging I kept getting back from a `docker push` command was also misleading:
> permission_denied: installation not allowed to Write organization package 

This was very confusing and had be quadruple checking app permissions, installs, and package settings to try and figure out what was preventing access. Looking online wasn't very helpful either as the GitHub docs indicate how to set up GitHub Actions access for this flow, but none explicitly mentioned that you couldn't use a GitHub App token so outside of GH actions.

This all felt pretty bad, and is probably going to result in me moving my images elsewhere.

**Reply by renan-alm** [reply]
@238alexg, we are all in the same boat when it comes to GH Apps, but a quick input on your issue: A Fine-Grained Token can be owned by an Organization (Classic PATs don't have that feature) which can be a little better than actual personal ones. Not sure you knew it but it might help your scenario.

**Reply by 238alexg** [reply]
Thanks @renan-alm! Not sure how I didn't know about organization fine grained tokens but that will work for my use case 😎 

I am still using a GitHub app for some other things and it would be nice to leverage the same credentials for this, but at least I can use your suggestion to solve my issue with packages 🙏

**Reply by thegrumpylion** [reply]
@renan-alm didn't know that, thanks for sharing. still, there is no packages permissions for fine-grained tokens.

**Reply by utkarsh-um** [reply]
@238alexg @thegrumpylion was either of you able to use "organization fine grained tokens" for packages?

**Comment by tinhlhm-fgs** [comment]
any update?

**Comment by andremarcelotanner-cohere** [comment]
Hello from 2026, still not supported :)

**Reply by joaosa** [reply]
and another hello :). having this would be great!

**Reply by celoprd** [reply]
Hello, crazy that after almost 6 years. Is it possible to ping someone from github on this ? I mean, mark the discussion as answered does not mean that problem is solved no ?

**Comment by Nikita-Guchakov** [comment]
Same here, still not supported and still needed. GITHUB_TOKEN is not an option here, because we need to provide CI access to another repository of our organization

**Comment by keidyz** [comment]
Funny how every 2 years or so, I always end up trying to use gh apps for packages but then get stunned when it doesn't work and then I end up remembering that it really isn't supported.

It's been 6 years now since this discussion opened- now I'm just stunned that it hasn't been fixed yet.

**Comment by CpuID** [comment]
This would be extremely useful to have implemented, pity its sitting with so much interest for so long :(

**Comment by Parth-Sabhadiya-Imprivata** [comment]
It would be great to have GH apps to use GH packages. I did all option try with GH app permission on the package, but it says denied
